### PR TITLE
Only essential media queries

### PR DESCRIPTION
**Pull request by Sam152**

### CODE REVIEW

**Comment by oliverklee** [comment]
Wow, this was fast! :heart:  I'll have a look at it later.

In the meantime, could you please squash your commits into one commit? Thanks!

**Comment by Sam152** [comment]
Squashed.

**Comment by oliverklee** [comment]
Could you please prefix the commit message subject line with "[FEATURE]" and remove the period at the end of the commit message subject line?

And could you please add "Fixes #202" to the commit message description so merging this PR will autoclose that issue?

File: Tests/Unit/EmogrifierTest.php [hunk 1]
```diff
@@ -1519,4 +1519,27 @@ public function emogrifyAutomaticallyClosesUnclosedTag()
             $this->subject->emogrify()
         );
     }
+
+    /**
+     * @test
+     */
+    public function irrelevantMediaQueriesAreRemoved()
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Please let's split this into two tests: one that tests that media queries with matches are not removed, and one that tests that media queries without matches are removed.

And could you also please also add one test for the case of empty media queries?

File: Tests/Unit/EmogrifierTest.php [hunk 2]
```diff
@@ -1108,7 +1108,7 @@ public function emogrifyWithValidMediaQueryContainsInnerCss($css)
     public function emogrifyForHtmlWithValidMediaQueryContainsInnerCss($css)
     {
         $html = $this->html5DocumentType . PHP_EOL . '<html><style type="text/css">' . $css
-            . '</style><h1></h1></html>';
+            . '</style><h1></h1><p></p></html>';
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Why this change?

File: Tests/Unit/EmogrifierTest.php [hunk 3]
```diff
@@ -1088,7 +1088,7 @@ public function validMediaPreserveDataProvider()
      */
     public function emogrifyWithValidMediaQueryContainsInnerCss($css)
     {
-        $html = $this->html5DocumentType . PHP_EOL . '<html><h1></h1></html>';
+        $html = $this->html5DocumentType . PHP_EOL . '<html><h1></h1><p></p></html>';
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Why this change?

File: Tests/Unit/EmogrifierTest.php (hunk 2 above)
Comment (Sam152) [inline Tests/Unit/EmogrifierTest.php]: Because the new behavior removes the media query CSS if it doesn't match any selectors. We need a <p></p> so the media queries stay in the document.

File: Classes/Emogrifier.php [hunk 4]
```diff
@@ -537,20 +548,36 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
     }
 
-
     /**
      * Copies the media part from CSS array parts to $xmlDocument.
      *
-     * @param string[] $cssParts
-     * @param \DOMDocument $xmlDocument
+     * @param string $mediaCss A string of CSS.
+     * @param \DOMDocument $xmlDocument The document to match against.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Could you please lowercase the first letter of the @param descriptions and remove the period (as the descriptions are no full sentences)?

File: Tests/Unit/EmogrifierTest.php (hunk 3 above)
Comment (Sam152) [inline Tests/Unit/EmogrifierTest.php]: Same as above.

File: Classes/Emogrifier.php [hunk 5]
```diff
@@ -537,20 +548,36 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
      *
      * @return void
      */
-    public function copyCssWithMediaToStyleNode(array $cssParts, \DOMDocument $xmlDocument)
+    public function copyCssWithMediaToStyleNode($mediaCss, \DOMDocument $xmlDocument)
     {
-        if (isset($cssParts['media']) && $cssParts['media'] !== '') {
-            $this->addStyleElementToDocument($xmlDocument, $cssParts['media']);
+        if (empty($mediaCss)) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This is a string. So the check should be !== '', not empty (as empty on strings will provide false positives).

File: Classes/Emogrifier.php [hunk 6]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
     }
 
     /**
+     * Parse a list of selectors from a string of css.
+     *
+     * @param string $css A string of raw CSS code.
+     *
+     * @return array
+     */
+    protected function parseSelectors($css)
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: private

File: Classes/Emogrifier.php [hunk 7]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
 
         return $xmlDocument->saveHTML();
     }
 
     /**
+     * Parse a list of selectors from a string of css.
+     *
+     * @param string $css A string of raw CSS code.
+     *
+     * @return array
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Let's make this more specific: string[]

File: Classes/Emogrifier.php [hunk 8]
```diff
@@ -537,20 +548,36 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
      * @return void
      */
-    public function copyCssWithMediaToStyleNode(array $cssParts, \DOMDocument $xmlDocument)
+    public function copyCssWithMediaToStyleNode($mediaCss, \DOMDocument $xmlDocument)
     {
-        if (isset($cssParts['media']) && $cssParts['media'] !== '') {
-            $this->addStyleElementToDocument($xmlDocument, $cssParts['media']);
+        if (empty($mediaCss)) {
+            return;
+        }
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please add a blank line after the guard clause.

**Comment by Sam152** [comment]
Feedback addressed.

File: Tests/Unit/EmogrifierTest.php (hunk 1 above)
Comment (Sam152) [inline Tests/Unit/EmogrifierTest.php]: Done.

File: Classes/Emogrifier.php [hunk 9]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
-        $this->copyCssWithMediaToStyleNode($cssParts, $xmlDocument);
+        $this->copyCssWithMediaToStyleNode($cssParts['media'], $xmlDocument);
 
         return $xmlDocument->saveHTML();
     }
 
     /**
+     * Parse a list of selectors from a string of css.
+     *
+     * @param string[] $css a string of raw CSS code
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: No, this is just a string. The return type should be a string[].

File: Classes/Emogrifier.php [hunk 10]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
         }
 
-        $this->copyCssWithMediaToStyleNode($cssParts, $xmlDocument);
+        $this->copyCssWithMediaToStyleNode($cssParts['media'], $xmlDocument);
 
         return $xmlDocument->saveHTML();
     }
 
     /**
+     * Parse a list of selectors from a string of css.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Generally, the first sentence in a method documentation should be third person (with "This method" omitted): "Parses a list …"

File: Tests/Unit/EmogrifierTest.php [hunk 11]
```diff
@@ -1519,4 +1519,57 @@ public function emogrifyAutomaticallyClosesUnclosedTag()
...
         );
     }
+
+    /**
+     * @test
+     */
+    public function irrelevantMediaQueriesAreRemoved()
+    {
+        $usefulQuery = '@media all and (max-width: 500px) { p { color:red; } }';
+        $uselessQuery = '@media all and (max-width: 500px) { em { color:red; } }';
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: For this test, $usefulQuery is not needed.

File: Tests/Unit/EmogrifierTest.php [hunk 12]
```diff
@@ -1519,4 +1519,57 @@ public function emogrifyAutomaticallyClosesUnclosedTag()
...
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function relevantMediaQueriesAreRetained()
+    {
+        $usefulQuery = '@media all and (max-width: 500px) { p { color:red; } }';
+        $uselessQuery = '@media all and (max-width: 500px) { em { color:red; } }';
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: For this test, $uselessQuery is not needed.

**Comment by oliverklee** [comment]
I've just merged three other pull requests. Could you please rebase?

File: Classes/Emogrifier.php [hunk 13]
```diff
@@ -226,44 +226,8 @@ public function emogrify()
...
-
-            // now sort the selectors by precedence
-            usort($allSelectors, array($this,'sortBySelectorPrecedence'));
-
-            $this->caches[self::CACHE_KEY_CSS][$cssKey] = $allSelectors;
-        }
-
-        foreach ($this->caches[self::CACHE_KEY_CSS][$cssKey] as $value) {
+        $selectors = $this->parseSelectors($cssParts['css']);
+        foreach ($selectors as $value) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: I propose changing $value to $selector (as $selectors contains selectors).

File: Classes/Emogrifier.php (hunk 10 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: css -> CSS

File: Classes/Emogrifier.php [hunk 14]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+     * Parse a list of selectors from a string of css.
+     *
+     * @param string[] $css a string of raw CSS code
+     *
+     * @return array
+     */
+    private function parseSelectors($css)
+    {
+        $cssKey = md5($css);
+        if (!isset($this->caches[self::CACHE_KEY_CSS][$cssKey])) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This condition has a quite long content. Please let's change it to a guard clause.

File: Classes/Emogrifier.php [hunk 15]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+            // process the CSS file for selectors and definitions
+            preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mis', $css, $matches, PREG_SET_ORDER);
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This comment is here because the preg match is not very clear. Please let's move it to a well-named private method instead.

File: Classes/Emogrifier.php [hunk 16]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+
+            $allSelectors = array();
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please add an @var for $allSelectors:

/** @var string[][] $allSelectors */

File: Classes/Emogrifier.php [hunk 17]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+            foreach ($matches as $key => $selectorString) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Maybe we can rename $matches to be more semantic concerning what is matched.

File: Classes/Emogrifier.php [hunk 18]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                // if there is a blank definition, skip
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This comment is not needed as the code already is "speaking" enough.

File: Classes/Emogrifier.php [hunk 19]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                if (!strlen(trim($selectorString[2]))) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please use !== '' instead of strlen as this makes the intent of the code more clear.

File: Classes/Emogrifier.php [hunk 20]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                    continue;
+                }
+
+                // else split by commas and duplicate attributes so we can sort by selector precedence
+                $selectors = explode(',', $selectorString[1]);
+                foreach ($selectors as $selector) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: I find that the naming $allSelectors and $selectors makes it hard to grok the difference. Let's use names that make the difference more clear.

File: Classes/Emogrifier.php [hunk 21]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+            preg_match_all('/(?:^|[\\s^{}]*)([^{]+){([^}]*)}/mis', $css, $matches, PREG_SET_ORDER);
+
+            $allSelectors = array();
+            foreach ($matches as $key => $selectorString) {
+                // if there is a blank definition, skip
+                if (!strlen(trim($selectorString[2]))) {
+                    continue;
+                }
+
+                // else split by commas and duplicate attributes so we can sort by selector precedence
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Let's make this another small, well-named method so the comment is not needed anymore.

File: Classes/Emogrifier.php [hunk 22]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                $selectors = explode(',', $selectorString[1]);
+                foreach ($selectors as $selector) {
+                    // don't process pseudo-elements and behavioral (dynamic) pseudo-classes;
+                    // only allow structural pseudo-classes
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Private method instead of the comment?

File: Classes/Emogrifier.php [hunk 23]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                    if (strpos($selector, ':') !== false && !preg_match('/:\\S+\\-(child|type)\\(/i', $selector)) {
+                        continue;
+                    }
+
+                    $allSelectors[] = array('selector' => trim($selector),
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please move the first array element to its own line, too.

File: Classes/Emogrifier.php [hunk 24]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                        'attributes' => trim($selectorString[2]),
+                        // keep track of where it appears in the file, since order is important
+                        'line' => $key,
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: If $key actually is the line (or the line number), please let's rename it.

File: Classes/Emogrifier.php [hunk 25]
```diff
@@ -296,12 +260,59 @@ public function emogrify()
...
+                    );
+                }
+            }
+
+            // now sort the selectors by precedence
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This comment is not needed as the code already is "speaking" enough.

File: Classes/Emogrifier.php [hunk 26]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
         return trim($style);
     }
 
-
     /**
      * Copies the media part from CSS array parts to $xmlDocument.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This comments needs to change to the method's new behavior.

File: Classes/Emogrifier.php [hunk 27]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
     {
-        if (isset($cssParts['media']) && $cssParts['media'] !== '') {
-            $this->addStyleElementToDocument($xmlDocument, $cssParts['media']);
+        if ($mediaCss === '') {
+            return;
+        }
+
+        $xpath = new \DOMXPath($xmlDocument);
+        $retainedMediaQueries = [];
+        // Split the CSS into individual media queries.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: -> well-named, private method

File: Classes/Emogrifier.php [hunk 28]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
+        preg_match_all('#(?<query>@media[^{]*\{(?<css>(.*?)\})(\s*)\})#', $mediaCss, $mediaQueries);
+        foreach ($mediaQueries['css'] as $key => $mediaQuery) {
+            // Check if any of the selectors inside the media query apply to the
+            // document.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: -> private method

File: Classes/Emogrifier.php [hunk 29]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
+            $selectors = $this->parseSelectors($mediaQuery);
+            foreach ($selectors as $selector) {
+                $nodesMatchingSelector = $xpath->query($this->translateCssToXpath($selector['selector']));
+                if ($nodesMatchingSelector && $nodesMatchingSelector->length !== 0) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: $nodesMatchingSelector is not always a boolean. So the first part of the condition needs to be $nodesMatchingSelector !== false.

File: Classes/Emogrifier.php [hunk 30]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
+                    $retainedMediaQueries[] = $mediaQueries['query'][$key];
+                }
+            }
         }
+        // Add only the media queries which have relevance for the document.
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: This comment should rather be part of the method comment, I think.

File: Classes/Emogrifier.php [hunk 31]
```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
...
+        $this->addStyleElementToDocument($xmlDocument, implode('', $retainedMediaQueries));
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: If $glue is an empty string, you can also use implode with a single parameter instead.

**Comment by Sam152** [comment]
Given there are a lots of small but opinionated changes here, do you want to take over the PR and make the stylistic changes on commit?

**Comment by oliverklee** [comment]
I could do the stylistic changes for you. And first, And before, I'll need you to make the changes that relate to structure and method/variable naming (as you know better than I do what the intention behind your changes is (and you probably understand the CSS parsing code better than I do)).

**Comment by oliverklee** [comment]
Please don't feel frustrated or discouraged due to my review comments. I want to make the code from your PR as readable and clear as possible, which in turn helps make Emogrifier more readable and maintainable.

I gather that you understand the way Emogrifier processes the CSS quite well (which I don't), so I really appreciate the way you PR can help make the corresponding part of the code better. :heart:

**Comment by oliverklee** [comment]
@Sam152 Could you please rebase this PR? It currently cannot be merged.

Background: The master on your fork https://github.com/Sam152/emogrifier is behind this repository's master: "This branch is 7 commits behind jjriv:master"

This is how you sync your for to this repository:

(once) add this repo as another remote:

git add remote upstream git@github.com:jjriv/emogrifier.git

every now an then (i.e., when the official repo has new commits):

git checkout master
git fetch upstream
git merge upstream/master

after that, for every of your PRs:
git checkout your-pr-branch
git rebase master
(then resolve any rebase conflicts)
git push -f

Here's more documentation on this:
https://help.github.com/articles/configuring-a-remote-for-a-fork/
https://help.github.com/articles/syncing-a-fork/
https://help.github.com/articles/resolving-merge-conflicts-after-a-git-rebase/

Hope this helps! :smile:

**Comment by oliverklee** [comment]
@Sam152 I've just merged a few other PRs. Could you please rebase and update your PR? Thanks!

**Comment by Sam152** [comment]
Rebased. Most of the comments are stylistic things against a portion of code I moved and did not write. Feel free to run with it from here.

**Comment by Sam152** [comment]
Any progress on this? I think refactoring the code which was not written but moved should be out of scope for this PR.

**Comment by oliverklee** [comment]
I'll re-review (and hopefully merge) this PR this week.

File: Classes/Emogrifier.php (hunk 29 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: done

File: Classes/Emogrifier.php (hunk 30 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: done

File: Classes/Emogrifier.php (hunk 31 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: done

File: Tests/Unit/EmogrifierTest.php (hunk 12 above)
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: done

File: Classes/Emogrifier.php (hunk 28 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: done

File: Classes/Emogrifier.php (hunk 27 above)
Comment (oliverklee) [inline Classes/Emogrifier.php]: done

**Comment by oliverklee** [comment]
Hi @Sam152, I've cleaned up and merged this PR as #217. Thanks for contributing! :heart:

**Comment by Sam152** [comment]
Thanks. Appreciate you taking it to a state of completion.
